Resolve "Metric Collector"

geunwoonoh opened the pull request:
As mentioned in https://github.com/alpacanetworks/alpamon/issues/9, we have added the following packages to implement a real-time monitoring system for AIP:

- `Collector`
   - This package creates and manages the following subcomponents required for metric collection:
       - `Check`
       - `Scheduler`
       - `Transporter`

- `Check`
   - The package functions as an individual metric monitoring unit.
   - It is responsible for metric collection and data management.
   
- `Scheduler`
   - This package manages the execution timing and frequency of Checks.
   
- `Transporter`
   - This package is responsible for sending collected metrics to the `alpacon-server`.
  
Closes https://github.com/alpacanetworks/alpamon/issues/9

royroyee commented:
It seems the pipeline error is due to the GitHub Action changes not being present in this branch. It would be good to merge the changes and then proceed.

geunwoonoh committed:
Add check components to Collector

A Check component, which acts as
an individual monitoring unit for
specific metrics, has been added to
Collector.
A Check component has been added
to Collector to perform individual
checks for specific metrics, such as
CPU usage, disk usage, disk I/O,
memory, and network traffic.

geunwoonoh committed:
Add Scheduler to Collector

Scheduler component has been
added to manage when and how
often Checks are executed.

geunwoonoh committed:
Add transporter to Collector

Transporter component has been
added to transmit the collected
metrics to the alpacon-server.

geunwoonoh committed:
Add collector for monitoring system

Collector component has been
added to collect real-time metrics
from server and transmit them to
alpacon-server.

geunwoonoh committed:
Add Post() for post request

Add Post() to make POST requests
using Session.

geunwoonoh committed:
Add logic to start collector

Add logic to start collector.

geunwoonoh committed:
Add disk usage related fields to CheckResult

Fields have been added to CheckResult due to
model changes in alpacon-server.

geunwoonoh committed:
Refactor to collect only physical interfaces

Refactored to collect only physical interfaces by
checking for MAC addresses.

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 9-metric-collector

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 9-metric-collector

geunwoonoh committed:
Add schemas

Decide to use SQLite to store real-time
metrics temporarily for batch processing.
Define schemas and add corresponding
ent code.
Add CPU, Memory, Traffic, DiskUsage,
DiskIO, and per-hour metrics for
CPUPerHour, MemoryPerHour,
TrafficPerHour, DiskUsagePerHour,
and DiskIOPerHour.

geunwoonoh committed:
Add ent package

Add ent package for quering data with
orm, ensuring type safety, and supporting
versioned migrations.

geunwoonoh committed:
Refactor Check

Add functions to store data in SQLite
database for batch processing.
Moved existing checks to realtime folder
to separate batch checks.

geunwoonoh committed:
Add hourly batch check

Add hourly batch checks for metrics.

geunwoonoh committed:
Add DB client connection logic.

Add DB client creation logic upon
alpamon startup.

geunwoonoh committed:
Update according to changes

Update to incorporate ent package
and additional batch checks.

geunwoonoh committed:
Add .gitignore

Add .gitignore  to prevent pushing code
generated by ent package.

geunwoonoh committed:
Update .goreleaser.yaml

Add a command to before.hooks to
generate ent-related code before building.

geunwoonoh committed:
Add save function to save hourly batch metric

Add save functions to store hourly batched metrics
for each metric.

geunwoonoh committed:
Refactor CheckError

Refactor CheckError to handle errors for
each query type as the number of query
types increased.
Update the existing realtime checks accordingly.

geunwoonoh committed:
Add daily batch checks

Add checks for batching each metric's
data daily.

geunwoonoh committed:
Refactor QuerySet

To share QuerySet structs defined for
fetching data from database in both
daily and hourly batch jobs, move them
to base package.
The hourly batch job code was modified accordingly.

geunwoonoh committed:
Minor fix

Fix typo

geunwoonoh committed:
Add delete query for old metrics

Add functions to each metric that
execute delete queries to remove
old data after batch processing.
Daily batch jobs delete data from
the previous day, while hourly batch
jobs remove data from the previous
hour.

geunwoonoh committed:
Add collectorArgs and collectConf type

Add collectConf type to facilitate the
retrieval of metric collection configuration
from alpacon-server.
Add collectorArgs type to encapsulate the
parameters required for collector instantiation.
The initialization process for the collector
has been streamlined into the
InitCollector() function.
Inside InitCollector(), fetch the metric
collection configuration and construct
the object to be passed as an argument.

geunwoonoh committed:
Add CheckArgs and CheckStrategy type

To promote code maintainability, add
CheckArgs type to hold the arguments
passed to Check constructor.
As part of the refactoring process, move
CheckStrategy interface from the check
package to this package to improve
cohesion and reduce dependencies within
CheckFactory.

geunwoonoh committed:
Refactor CreateCheck()

Initial approach involved using switch
statement to invoke different NewCheck()
 depending on CheckType, leading to a
less maintainable codebase due to code
redundancy.
To improve code organization and
maintainability, standardize the creation
of Check by adding newCheck() function
type and refactoring the various
package-specific NewCheck() functions
to conform to this type.
This unification has enabled to simplify
the logic within CreateCheck() of
CheckFactory by establishing a clear mapping between CheckType and
the corresponding newCheck() function.

geunwoonoh committed:
Update due to apply changes

Update code to reflect changes
mentioned in the previous commit.

geunwoonoh committed:
Add cleanup check

Add cleanup check to delete all data
in the database daily.

geunwoonoh committed:
Adjust buffer size and waitgroup number

Adjust buffer size and waitgroup count to match
the number of checks received from alpacon-server.

geunwoonoh committed:
Add types.go in base package

To improve code modularity and maintainability,
move all check-related type definitions from base.go
to a new file, types.go.

geunwoonoh committed:
Delete running flag in scheduler

Due to the idempotency of task execution,
delete the unnecessary running flag.

geunwoonoh committed:
Add retry logic to checks

Add retry mechanism for checks that fail.
Use exponential backoff for retrying failed checks.

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 9-metric-collector

geunwoonoh committed:
Update due to changes in alpacon-server

Due to changes in Traffic-related models
in alpacon-server, update the system to
receive input/output data in bps and pps
format, rather than raw bytes and packets.

geunwoonoh committed:
Add functions to calculate bps/pps

Add utility functions to calculate bps
and pps.

geunwoonoh committed:
Minor fix

To omit unnecessary file systems from
the collection process on macOS,
fix excludedFileSystems.
Upon its initial data collection, alpamon
is now configured to set the initial disk
read and write bytes to 0

geunwoonoh committed:
Refactor Scheduler

Add retry logic to Scheduler to handle errors
that may occur during check execution.
To improve performance, add worker pool and
task queue for parallel task execution.
To optimize performance and simplify concurrent
access, replace the existing data structure with
sync.Map for task registration and locking.

geunwoonoh committed:
Refactor Checks

To enforce the single responsibility principle
and confine the Check's duties to metric tasks,
retry logic has been moved to the Scheduler.
To support database rollbacks in check execution
failures, add explicit transaction management.

geunwoonoh committed:
Update due to apply changes

Update due to follow a refactoring of
Checks and Scheduler.

geunwoonoh committed:
Refactor Transporter

Refactor Transporter that was violating
Dependency Inversion Principle.
Update Collector due to apply Transporter's
change.

geunwoonoh committed:
Add retry logic at Transporter & Collector

Refactor Collector and Transporte to
retry with exponential backoff in Collector
when data transmission to Alpacon Server
via Transporter's Send() fails.

geunwoonoh committed:
Apply versioned migrations

Apply versioned migration to ensure
stability when adding new metrics or
expanding the system in the future.

geunwoonoh committed:
Add ariga.io/atlas-go-sdk package

Add ariga.io/atlas-go-sdk package
to support versioned migration.

geunwoonoh committed:
Add InitDB()

Add InitDB() to execute both
the DB client acquisition logic
and DB migration logic at once,
which were previously executed
separately for readability.

geunwoonoh committed:
Minor fix

Fix .gitignore
Fix typo in pkg/collector/collector.go

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 9-metric-collector

geunwoonoh committed:
Fix SQL query for disk usage

A defect in the select query of hourly disk usage
batch job check was causing incorrect data retrieval.
To fix this, the query was refactored to employ
a subquery, ensuring accurate data extraction.
DiskUsageQuerySet has been correspondingly
adjusted.

geunwoonoh committed:
Delete mount_point field

Delete the unnecessary mount_point field
from diskusageperhour table
Fix the corresponding migration file.

geunwoonoh committed:
Add --feature flag at ent command

To support sql/modifier feature in ent package,
add --feature flag and the corresponding functionality
to command

geunwoonoh committed:
Minor fix

Fix retry logic to ensure that subsequent
tasks continue to execute even if a retry fails.

geunwoonoh committed:
Add logger

Add logger for debugging purposes
in case of errors.

geunwoonoh committed:
Add ariga.io/atlas-go-sdk package

Add ariga.io/atlas-go-sdk package for
versioned migrations.

geunwoonoh committed:
Refactor for building Linux packages

Add Atlas CLI installation script to postinstall.sh
for Linux package builds.
Refactor .goreleaser.yaml to generate ent-related code.
Add the database file path to tmpfile.conf.

geunwoonoh committed:
Update README.md

Add explanations on how to generate
ent-related code.

geunwoonoh committed:
Fix migration directory path

A problem occurred where the migration directory
was specified with a relative path, making it
impossible for the binary file to locate this path.
To resolve this issue, embed package was used to
specify migration directory path.

geunwoonoh committed:
Add logger

Add logger for debugging purposes
in case of errors.

geunwoonoh committed:
Fix builds env

Fix .goreleaser.yaml to enable cgo for using
go-sqlite3 package.

geunwoonoh committed:
Delete 	mattn/go-sqlite3 package

Remove mattn/go-sqlite3 package to
avoid using cgo.

geunwoonoh committed:
Update due to apply changes

Update to reflect the removal of
mattn/go-sqlite3 package.

geunwoonoh committed:
Minor fix

Undo CGO_ENABLED

geunwoonoh committed:
Update github action yml

Update github action yml due to
apply changes

geunwoonoh committed:
Update github action yml

Update github action yml due to
apply changes

geunwoonoh committed:
Minor fix

Apply go-lint.

royroyee commented on cmd/alpamon/command/root.go:
```diff
@@ -61,6 +64,25 @@ func runAgent() {
+	defer cancel()
+
+	// DB
+	client := db.InitDB(ctx)
```
- The `collector` variable name matches the package name, which can cause confusion.
- `err :=` pattern directly for simplicity, especially in shorter code blocks
- The error handling goroutine appears unnecessary and may introduce complexity. Simplify error handling without spawning a separate goroutine.
- `context` might not be necessary here, as it’s not being effectively utilized. Remove it if not needed or refactor to use it meaningfully

geunwoonoh replied:
I have applied your feedback. Please review and resolve it:
- The context initialized in root.go has been modified to include a timeout within the InitDB() function. For detailed changes, please refer to the provided link.
   - https://github.com/alpacanetworks/alpamon/pull/21/commits/94ba7e5d42af1153056e09e15880df025f9aaf25

royroyee replied:
The for range loop is blocking and will keep waiting until it receives an error from the channel.

In root.go, it would be better to keep the code concise and focused only on initialization logic.

Also, since other packages in root.go do not pass context, it might be better to declare and handle the context within the DB logic itself, if it is necessary, rather than declaring it in root.go.

geunwoonoh replied:
The code has been modified to encapsulate context declaration within InitDB() function and to localize context declaration for the collector within its specific logic.

royroyee commented on pkg/collector/collector.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+func NewCollector(args collectorArgs) (*Collector, error) {
+	transporter, err := args.transportFactory.CreateTransporter(args.session)
```
Many variables share names with their packages, causing confusion. Please use distinct names for clarity.

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

royroyee commented on pkg/collector/scheduler/scheduler.go:
```diff
@@ -0,0 +1,132 @@
+		task.nextRun = time.Now().Add(task.interval)
+	}()
+
+	for attempt := 0; attempt <= s.retryConf.MaxRetries; attempt++ {
```
Please refer to the existing backoff and related implementations in Alpamon for improving readability and simplifying the retry logic.

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

royroyee commented on pkg/db/client.go:
```diff
@@ -0,0 +1,35 @@
+	return client, initErr
+}
+
+func Close() error {
```
This function is currently unused.

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

royroyee commented on pkg/db/client.go:
```diff
@@ -0,0 +1,35 @@
+	_ "github.com/glebarez/go-sqlite"
+)
+
+var (
```
Please avoid using global variables whenever possible.

geunwoonoh replied:
I have added a DBClientManager to maintain the client using the singleton pattern and applied the feedback.

royroyee commented on pkg/db/migrate.go:
```diff
@@ -0,0 +1,61 @@
+		log.Error().Err(err).Msgf("failed to open migration dir: %v", err)
+		return err
+	}
+	defer workDir.Close()
```
Missing error handling in defer statement.

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

geunwoonoh committed:
Refactor realtime NET & DISK_IO check

Refactor Checks that collect, store, and transmit
real-time network traffic and disk I/O to ensure
collected data is meaningful.
Fix disk I/O check to collect bytes per second (BPS)
to measure speed, whereas it previously only collected
disk read/write bytes.
Due to the long collection interval, the data collected
by the existing traffic check for input/output bps and pps
was not meaningful.
To improve data accuracy and timeliness, refactor the
check to separate the collection and storage processes
from the data transmission process.

geunwoonoh committed:
Update to apply the changes

Update due to changes in alpacon-server
and refactoring of alpamon.

royroyee commented on pkg/collector/collector.go:
```diff
@@ -0,0 +1,228 @@
+	return nil
+}
+
+func (c *Collector) Start(ctx context.Context) error {
```
The Start function always returns nil, but root.go is handling its error return. 

Consider revising the error handling logic.

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

royroyee commented on pkg/collector/collector.go:
```diff
@@ -0,0 +1,228 @@
+			return
+		case <-c.stopChan:
+			return
+		case metric := <-c.buffer.SuccessQueue:
```
Nested select reduces readability; consider refactoring to simplify the logic

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

royroyee commented on pkg/collector/transporter/transporter.go:
```diff
@@ -0,0 +1,59 @@
+		return err
+	}
+
+	if statusCode > 300 {
```
Use explicit status code checks for consistency

Our project already follows this approach in files like reporter.go and commit.go.

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

royroyee commented on pkg/collector/scheduler/scheduler.go:
```diff
@@ -0,0 +1,132 @@
+
+const (
+	MAX_RETRIES     int           = 5
+	MAX_RETRY_TIMES time.Duration = 1 * time.Minute
```
Type time.Duration can be omitted for simplicity.

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

royroyee commented on pkg/collector/check/batch/hourly/cpu/cpu.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+func (c *Check) queryCPUUsage(ctx context.Context) (base.MetricData, error) {
+	queryset, err := c.getCPU(ctx)
```
Use camel case for variable naming, and apply this change consistently throughout this pr (queryset, .. etc)

geunwoonoh replied:
I have applied your feedback. Please review and resolve it.

royroyee reviewed (commented):
There seem to be many redundant parts in the code. While some may be unavoidable, it would be great to minimize them or improve them over time. 

For code style, please refer to the existing implementation as much as possible for consistency. 

Since there are multiple contexts and channels in use, please test various cases and consider using pprof to check for potential goroutine leaks

geunwoonoh committed:
Minor fix

Apply go code convention.
Apply code review

geunwoonoh committed:
Fix retry logic in collector's scheduler

To simplify and improve readability of
the retry logic with exponential backoff,
RetryStatus has been added to ScheduledTask.

geunwoonoh committed:
Add DBClientManager

To manage the database client as a singleton,
add DBClientManager.

geunwoonoh committed:
Add test cases for checks

 To ensure reliability and maintainability,
 add test cases for checks.

geunwoonoh committed:
Add context for migration

The context used in the migration process
has been updated to limit 5, superseding
the context originally defined in root.go.

royroyee committed:
Minor fix

royroyee committed:
Minor fix

royroyee commented on pkg/collector/collector.go:
```diff
@@ -0,0 +1,221 @@
+	return fmt.Errorf("max retries exceeded for metric %s", metric.Type)
+}
+
+func (c *Collector) Stop() {
```
Is this method currently in use? I couldn’t find where it is being called.

Also, when working with goroutines and channels, **panic can occur if a closed channel is accessed**. Since this PR involves multiple channels, please pay extra attention to avoid panic.

geunwoonoh replied:
I've added a call to `defer metricCollector.Stop()` in `root.go`:
- https://github.com/alpacanetworks/alpamon/pull/21/commits/1cb6a8f1016246883a67124b3b55ab84813e87f7#diff-3c43b14aa7f9f7a6c37f5fb29c403060d61545082c4972bc928f29f6d9ca29a6R72
Additionally, i've refactored the code to use the "comma-ok" idiom for checking if a channel is closed when reading data:
- https://github.com/alpacanetworks/alpamon/pull/21/commits/1cb6a8f1016246883a67124b3b55ab84813e87f7#diff-82dc411a0ed86675e9efddaa43dadc4fc6a75ea0e672400429e8ea39424da44cR162
- https://github.com/alpacanetworks/alpamon/pull/21/commits/1cb6a8f1016246883a67124b3b55ab84813e87f7#diff-82dc411a0ed86675e9efddaa43dadc4fc6a75ea0e672400429e8ea39424da44cR186

royroyee committed:
Minor fix

geunwoonoh committed:
Refactor declaration location

Refactor the code to move the context declaration
for database migrations outside of root.go, allowing
root.go to focus solely on initialization tasks.

geunwoonoh committed:
Refactor collector initialization logic

Refactor the code to move the context declaration
for collector outside of root.go, allowing
root.go to focus solely on initialization tasks.
Refactor collector to utilize the "comma-ok" idiom
for managing goroutines.

geunwoonoh committed:
Merge remote-tracking branch 'origin/9-metric-collector' into 9-metric-collector

geunwoonoh committed:
Rename due to changes in alpacon-server

Rename due to changes in alpacon-server.

geunwoonoh committed:
Add step for install atlas cli

Add a step to install Atlas CLI for
running check-related tests.

geunwoonoh committed:
Minor fix

Fix typo.

geunwoonoh committed:
Add devfs to excludedFileSystems

Add devfs to excludedFileSystems as it's a virtual
filesystem and doesn't require separate collection.

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 9-metric-collector

geunwoonoh committed:
Add script to delete db file

Add script to delete db file.
